src/modules/audio/index.ts

```diff
@@ -5,4 +5,196 @@
+        return inputsList;
+    },
+
+    /**
+     * Set the list of known audio input devices. This remembers the list locally and
+     * also updates the Vuex store with the new information.
+     *
+     * @param {MediaDeviceInfo[]} pAudioInputs The list of audioinput devices to remember
+     */
+    saveInputsList(pAudioInputs: MediaDeviceInfo[]): void {
+        Audio._inputsList = pAudioInputs;
+
+        Store.commit(Mutations.MUTATE, {
+            property: "audio.inputsList",
+            value: pAudioInputs
+        });
+    },
+
+    async requestInputAccess(): Promise<Nullable<MediaStream>> {
+        if (Audio._stream) {
+            return Audio._stream;
+        }
+        const stream = await navigator.mediaDevices.getUserMedia({ audio: true, video: false });
+        if (stream) {
+            await Audio.handleRequestInputSuccess(stream);
+            return Audio._stream;
+        }
+        return null;
+    },
+
+    /**
+     * Stop all audio activities on the stream.
+     */
+    stopCurrentInputStream(): void {
+        if (Audio._stream) {
+            Audio._stream.getTracks().forEach((track) => {
+                track.stop();
+            });
+        }
+        Audio.revokeCaptureAccess();
+    },
+
+    /**
+     * Set the state of the audio device to disconnected and unallocated.
+     */
+    revokeCaptureAccess(): void {
+        Audio._hasInputAccess = false;
+        Audio._currentInputDevice = <MediaDeviceInfo><unknown>undefined;
+        Audio._stream = undefined;
+
+        Store.commit(Mutations.MUTATE, {
+            property: "audio",
+            with: {
+                connected: false,
+                currentInputDevice: "UNKNOWN",
+                hasInputAccess: false
+            }
+        });
+    },
+
+    /**
+     * Request access to a specific media device.
+     *
+     * This requests the media device from the browser and, if connected, updates the
+     * Vuex state and sets up the device for operation.
+     *
+     * If a stream is already connected, this function first disconnects it before requesting.
+     * the new device.
+     *
+     * There is some protection from being called twice while already waiting for a device.
+     *
+     * @param requestedDeviceId the deviceId from the MediaDeviceInfo structure of device to request access
+     * @returns {Promise<Nullable<MediaStream>>} the connected media stream or "null" if access refused
+     */
+    async requestSpecificInputAccess(requestedDeviceId: string): Promise<Nullable<MediaStream>> {
+        if (Audio._stream) {
+            Audio.stopCurrentInputStream();
+        }
+
+        if (Audio._awaitingCapturePermissions === true) {
+            Log.error(
+                Log.types.AUDIO,
+                `Failed to request specific input device ID ${requestedDeviceId}`
+                 + ` due to an already awaiting request for input capture.`
+            );
+            return null;
```

Comment:
The function returns a Promise<boolean>. The boolean is 'true' on a successful login.

Reply:
My previous comment was all wrong -- I must have been looking at the wrong file.

In audio/index.ts, that routine returns a Nullable<MediaStream> so, if it allocates a media stream, that stream is returned. If there is some error, it returns 'null'.

This particular error is when requesting a specific device  and that device is awaiting capture permissions. The existing code returns an error (null) in this case. The code might need are rework so the awaiting is handled properly.  Maybe the Promise shouldn't be resolved until capture permissions have been granted.

This code will be reworked over the next few days so I'll have to consider that problem.

Reply:
All these routines have been moved around in PR#23 (moved into Audio.vue since it is all related to the UI and browser. This has changed error handling a little bit.